Implement `tbot init` subcommand and ACL management

timothyb89 opened the pull request:
This adds a new CLI subcommand to initialize a tbot destination directory by creating required files ahead of time and assigning proper permissions (and ACLs, where possible). It also adds a number of UX improvements as a result of some significant identity refactoring.

ACLs are used to allow the bot and end-user workflow to run as separate UNIX users. The bot can be assigned a dedicated UNIX user and given exclusive access to `/var/lib/teleport/bot` where the renewable certificate is stored. The end user then uses `tbot init` to create the secondary (non-renewable) certificate directory with only their own UNIX permissions, but uses ACLs to allow the bot user to deposit the non-renewable certificates to their own directory securely. Should an attacker compromise the workflow user, they may be able to steal the secondary certificates, but should not be able to access the bot's renewable certificates. This should limit the longevity and (depending on configuration) degree of access an attacker might gain on a compromised system.

## Summary of changes

1. Identity refactoring: save and load now accept kind tags and only attempt to store/load artifacts matching the specified tags
2. Implement `kinds` in the config which was previously not honored, so now only artifacts of the given kind are stored and loaded (i.e. ssh, tls, or both)
   * Also includes compatibility fixes to support one identity type (SSH or TLS) where previously quite a lot of code assumed a fully realized identity always existed.
3. Implements `tbot init` subcommand to configure file ACLs automatically
4. Add configurable ACL checking and symlink attack mitigation
   * ACL checking is warn-only by default on the bot's end, but can be configured to quit on error by setting `acls: on` in the directory YAML config.
   * Symlink attack mitigation makes writes fail if a path contains any symlinks. It's enabled by default on Linux 5.6+, but can be disabled by setting `symlinks: insecure` in the directory YAML config.
5. Test destinations before attempting a renewal.
   * Previously, if certificates couldn't be written, subsequent renewals would trigger a generation counter lock. The bot now attempts a write first to help ensure the destination is usable, and if not, fails early without locking the bot user.
6. Store a hash of the join token and throw away the old identity when it changes, fixing a common edge case
7. Create files directories automatically if they don't exist (to support the non-ACL use-case)
8. Add hint to use `tbot init` as part of the `tctl bots add ...` help message

## Usage

For simple deployments where file permissions aren't critical, the standard tbot instructions still work, i.e. just running the example command printed by `tctl bots add ...`. For users that wish to run the bot as a separate Linux user and restrict other users on the system from accessing the renewable certs, they can do the following (assuming their OS and filesystem support ACLs):

```bash
# on the cluster: create the bot user
$ tctl -c ~/.local/share/teledev/foo.example.com/teleport.yaml bots add test --roles=access
The bot token: 7ff4407e924cec953f679be719a93fdb
This token will expire in 59 minutes.

Run this on the new bot node to join the cluster:

> tbot start \
   --destination-dir=./tbot-user \
   --token=7ff4407e924cec953f679be719a93fdb \
   --ca-pin=sha256:634470e113d96a18ec1d76057fff0921f43347397fc107ccabe597dbbe07fdc8 \
   --auth-server=auth.example.com

Please note:

  - This invitation token will expire in 59 minutes
  - 192.168.122.10:3025 must be reachable from the new node

# on the bot node: create a user and storage dir for the bot
alice $ sudo useradd -r tbot
alice $ sudo mkdir -p /var/lib/teleport/bot
alice $ sudo chown tbot:tbot /var/lib/teleport/bot

# on the bot node as root, initialize the destination and grant read/write to the bot and end user
alice $ sudo tbot init --destination-dir=./tbot-user --bot-user=tbot --reader-user=alice

# now that permissions have been configured, we can start the bot like usual
# on the bot node as the bot user (using sudo for simplicity)
alice $ sudo -u tbot tbot start \
   --destination-dir=./tbot-user \
   --token=7ff4407e924cec953f679be719a93fdb \
   --ca-pin=sha256:634470e113d96a18ec1d76057fff0921f43347397fc107ccabe597dbbe07fdc8 \
   --auth-server=auth.example.com

# now certificates have been issued to ./tbot-user: owned by nobody, but written by tbot and readable by ailce

```

timothyb89 committed:
Add certificate renewal bot

This adds a new `tbot` tool to continuously renew a set of
certificates after registering with a Teleport cluster using a
similar process to standard node joining.

This makes some modifications to user certificate generation to allow
for certificates that can be renewed beyond their original TTL, and
exposes new gRPC endpoints:
 * `CreateBotJoinToken` creates a join token for a bot user
 * `GenerateInitialRenewableUserCerts` exchanges a token for a set of
   certificates with a new `renewable` flag set

A new `tctl` command, `tctl bots add`, creates a bot user and calls
`CreateBotJoinToken` to issue a token. A bot instance can then be
started using a provided command.

timothyb89 committed:
Cert bot refactoring pass

* Use role requests to split renewable certs from end-user certs
* Add bot configuration file
* Use `teleport.dev/bot` label
* Remove `impersonator` flag on initial bot certs
* Remove unnecessary `renew` package
* Misc other cleanup

timothyb89 committed:
Do not pass through `renewable` flag when role requests are set

This adds additional restrictions on when a certificate's `renewable`
flag is carried over to a new certificate. In particular, it now also
denies the flag when either role requests are present, or the
`disallowReissue` flag has been previously set.

In practice `disallow-reissue` would have prevented any undesired
behavior but this improves consistency and resolves a TODO.

timothyb89 committed:
Various tbot UX improvements; render SSH config

* Fully flesh out config template rendering
* Fix rendering for SSH configuration templates
* Added `String()` impls for destination types
* Improve certificate renewal logging; show more detail
* Properly fall back to default (all) roles
* Add mode hints for files
* Add/update copyright headers

timothyb89 committed:
Add stubs for tbot init and watch commands

timothyb89 committed:
Add gRPC endpoints for managing bots

* Add `CreateBot`, `DeleteBot`, and `GetBotUsers` gRPC endpoints
* Replace `tctl bot (add|rm|ls)` implementations with gRPC calls
* Define a few new constants, `DefaultBotJoinTTL`, `BotLabel`,
  `BotGenerationLabel`

timothyb89 committed:
Fix outdated destination flag in example tbot command

timothyb89 committed:
Bugfix pass for demo

* Fixed a few nil pointer derefs when using config from CLI args
* Properly create destination if `--destination-dir` flag is used
* Remove improper default on CLI flag
* `DestinationConfig` is now a list of pointers

timothyb89 committed:
Address first wave of review feedback

Fixes the majority of smaller issues caught by reviewers, thanks all!

timothyb89 committed:
Add doc comments for bot.go functions

timothyb89 committed:
Return the token TTL from CreateBot

timothyb89 committed:
Split initial user cert issuance from `generateUserCerts()`

Issuing initial renewable certificate ended up requiring a lot of
hacks to skip checks that prevented anonymous bots from getting
certs even though we'd verified their identity elsewhere (via token).

This reverts all those hacks and splits initial bot cert logic into a
dedicated `generateInitialRenewableUserCerts()` function which should
make the whole process much easier to follow.

timothyb89 committed:
Set bot traits to silence log messages

timothyb89 committed:
tbot log message consistency pass

timothyb89 committed:
Implement `tbot init` subcommand

This adds a new CLI subcommand to initialize a tbot destination
directory by creating required files ahead of time and assigning
proper permissions (and ACLs, where possible).

timothyb89 committed:
Resolve lints

timothyb89 committed:
Add config tests

timothyb89 committed:
Remove CreateBotJoinToken endpoint

Users should instead use the CreateBot/DeleteBot endpoints.

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/tbot

timothyb89 committed:
Create a fresh private key for every impersonated identity renewal

timothyb89 committed:
Hide `config` subcommand

timothyb89 committed:
Rename bot label prefix to `teleport.internal/`

timothyb89 committed:
Merge branch 'timothyb89/tbot' into timothyb89/tbot-init

timothyb89 committed:
Use types.NewRole() to create bot roles

timothyb89 committed:
Clean up error handling in custom YAML unmarshallers

Also, add notes about the supported YAML shapes.

timothyb89 committed:
Fetch proxy host via gRPC Ping() instead of GetProxies()

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/tbot

timothyb89 committed:
Update lib/auth/bot.go

Co-authored-by: Zac Bergquist <zmb3@users.noreply.github.com>

timothyb89 committed:
Fix some review comments

timothyb89 committed:
Add renewable certificate generation checks (#10098)

* Add renewable certificate generation checks

This adds a new validation check for renewable certificates that
maintains a renewal counter as both a certificate extension and a
user label. This counter is used to ensure only a single certificate
lineage can exist: for example, if a renewable certificate is stolen,
only one copy of the certificate can be renewed as the generation
counter will not match

When renewing a certificate, first the generation counter presented
by the user (via their TLS identity) is compared to a value stored
with the associated user (in a new `teleport.dev/bot-generation`
label field). If they aren't equal, the renewal attempt fails.
Otherwise, the generation counter is incremented by 1, stored to the
database using a `CompareAndSwap()` to ensure atomicity, and set on
the generated certificate for use in future renewals.

* Add unit tests for the generation counter

This adds new unit tests to exercise the generation counter checks.

Additionally, it fixes two other renewable cert tests that were
failing.

* Remove certRequestGeneration() function

* Emit audit event when cert generations don't match

* Fully implement `tctl bots lock`

* Show bot name in `tctl bots ls`

* Lock bots when a cert generation mismatch is found

* Make CompareFailed respones from validateGenerationLabel() more actionable

* Update lib/services/local/users.go

Co-authored-by: Nic Klaassen <nic@goteleport.com>

* Backend changes for tbot IoT and AWS joining (#10360)

* backend changes

* add token permission check

* pass ctx from caller

Co-authored-by: Roman Tkachenko <roman@goteleport.com>

* fix comment typo

Co-authored-by: Roman Tkachenko <roman@goteleport.com>

* use UserMetadata instead of Identity in RenewableCertificateGenerationMismatch event

* Client changes for tbot IoT joining (#10397)

* client changes

* delete replaced APIs

* delete unused tbot/auth.go

* add license header

* don't unecessarily fetch host CA

* log fixes

* s/tunnelling/tunneling/

Co-authored-by: Zac Bergquist <zmb3@users.noreply.github.com>

* auth server addresses may be proxies

Co-authored-by: Zac Bergquist <zmb3@users.noreply.github.com>

* comment typo fix

Co-authored-by: Zac Bergquist <zmb3@users.noreply.github.com>

* move *Server methods out of auth_with_roles.go (#10416)

Co-authored-by: Tim Buckley <tim@goteleport.com>

Co-authored-by: Zac Bergquist <zmb3@users.noreply.github.com>
Co-authored-by: Tim Buckley <tim@goteleport.com>

Co-authored-by: Roman Tkachenko <roman@goteleport.com>
Co-authored-by: Tim Buckley <tim@goteleport.com>
Co-authored-by: Zac Bergquist <zmb3@users.noreply.github.com>

Co-authored-by: Nic Klaassen <nic@goteleport.com>
Co-authored-by: Roman Tkachenko <roman@goteleport.com>
Co-authored-by: Zac Bergquist <zmb3@users.noreply.github.com>

timothyb89 committed:
Address another batch of review feedback

timothyb89 committed:
Addres another batch of review feedback

Add `Role.SetMetadata()`, simplify more `trace.WrapWithMessage()`
calls, clear some TODOs and lints, and address other misc feedback
items.

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/tbot

timothyb89 committed:
Fix lint

timothyb89 committed:
Add missing doc comments to SaveIdentity / LoadIdentity

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/tbot

timothyb89 committed:
Remove pam tag from tbot build

timothyb89 committed:
Update note about bot lock deletion

timothyb89 committed:
Another pass of review feedback

Ensure all requestable roles exist when creating a bot, adjust the
default renewable cert TTL down to 1 hour, and check types during
`CompareAndSwapUser()`

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/tbot

timothyb89 committed:
Merge branch 'timothyb89/tbot' into timothyb89/tbot-init

Merge note: BotKinds() now includes SSH certificates to support IoT
joining.

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/tbot-init

timothyb89 committed:
Remove ModeHint

timothyb89 committed:
Rename Identity.Cert and Identity.XCert

timothyb89 committed:
Add `symlinks` flag to tbot config

The optional symlinks flag for directory destinations allows users to
opt in / out of whichever symlink attack hardening mode is selected
by default.

timothyb89 committed:
Add mostly-working secure implementation of botfs.Create/Write

This adds symlink mode selection (secure, try-secure, insecure) and
Linux `Create()`/`Write()` implementations to open files safely.

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/tbot-init

timothyb89 committed:
Add configurable ACL modes and verify ACL support in tbot init

timothyb89 committed:
Initialize destinations at startup and test before renewal

This initializes destinations at startup (to create directories if
not using `tbot init`) and tests them to ensure the bot can write
_before_ attempting to renew certificates; this should prevent most
accidental generation counter locks.

timothyb89 committed:
Hide watch for now

timothyb89 committed:
Issue a new identity if a token change is detected

timothyb89 committed:
Warn if identity appears to be expired on startup

timothyb89 committed:
Fully implement ACL Verify and Configure

 - Fully implements ACL support for Linux
 - Adds bot-side verification support to ensure ACLs are configured
   properly at runtime.
 - Gracefully falls back to no ACLs if the platform / filesystem
   doesn't support them
 - Clear up outstanding lints

timothyb89 committed:
Make `tbot init` work without a config file

timothyb89 committed:
Show init instructions in tctl bots add

Also:
 - Make --bot-user a flag in init (the tctl instructions were
   confusing otherwise)
 - Handle IsOwnedBy sanely on unsupported platforms
 - Add Bold colorizing support

timothyb89 committed:
Clear some TODOs and rephrase tctl help

timothyb89 commented:
Screenshot of the revised help text for `tctl bots add ...`:
![image](https://user-images.githubusercontent.com/3459954/156480547-a3d1f5e7-d5a0-4635-b9e0-b13115cd50c1.png)

timothyb89 committed:
Fix typo

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/tbot-init

timothyb89 committed:
Fix token hash detection bug

timothyb89 committed:
Actually read and write certs with symlink enforcement

Also, fix a config loading bug where CheckAndSetDefaults() wasn't
being called in all cases with CLI destinations.

timothyb89 commented:
I could've sworn I tested this successfully, but SSH barfs because the ACL mask is stored in the group bits and SSH is (at least for our use-case) overly paranoid about keyfile permissions. I think `tbot init` instead needs to be run as root and create the destination as another user (e.g. `nobody`) then provide ACL entries for _both_ the bot _and_ the reader user, as this bypasses OpenSSH's permissions check.

timothyb89 committed:
Add workaround for OpenSSH permissions check with ACLs

OpenSSH has an overly-paranoid permissions check that forces key
files to be exclusively owner-readable. Unfortunately, for POSIX
compatibility purposes, when ACLs are set, the ACL mask is set as
the group permissions. This effectively makes any ACL incompatible
with OpenSSH.

However, OpenSSH's check does have an escape hatch: it only applies
if the current user is the owner of the file. Therefore, this change
tweaks the `tbot init` flow to create files as root, owned by a
separate user (either `nobody` or even the bot user), with ACL
permissions granting both the bot and reader user access to the
certificates. This effectively bypasses OpenSSH's permissions check
and should preserve our security boundaries.

timothyb89 committed:
Fix lints

timothyb89 committed:
Fix an improper directory chmod to 0600 if ACL test fails

timothyb89 committed:
First pass of tbot init unit tests

timothyb89 committed:
Add symlink tests and fix bug with resolving the default owner

timothyb89 committed:
Fix err misuse

timothyb89 committed:
Fix an ACL error if the bot or reader user is the owner.

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/tbot-init

timothyb89 committed:
Fix typo

timothyb89 commented:
I've now reworked the flow per the above comment. `tbot init` is now run by a (ideally) secondary user or root and grants permissions to the reader user via ACLs, which bypasses OpenSSH's permissions check. There's a few valid configurations for this:
* Run as root, owned by `nobody:nobody` (or any other permissionless user), with read/write granted via ACL to the bot and reader. `tbot init` will correct ownership and permissions automatically. This is the default.
* Run as the file owner, with separate bot and reader users.
* Run as the bot, with a separate reader user.
* These do work but generate a warning as they will break OpenSSH:
   * Bot, reader, and owner are all the same. This is allowed but weird, and will break OpenSSH. ACLs should just not be used (the warnings say to turn them off), and for that matter `tbot init` is superfluous.
   * Reader and owner are the same. This is the original behavior and, while valid for some use-cases, breaks OpenSSH.

zmb3 commented on tool/tbot/botfs/fs_linux.go:
```diff
@@ -0,0 +1,454 @@
+
+	data := make([]byte, 0, size)
+	for {
+		if len(data) >= cap(data) {
```
This is a bit confusing, as Go won't allow `len > cap`. What are you actually trying to accomplish here?

zmb3 commented on tool/tbot/botfs/fs_linux.go:
```diff
@@ -0,0 +1,454 @@
+	defer file.Close()
+
+	// Blatantly stolen from go's os/file.go since there's no stdlib function
+	// for reading everything from an _already open_ file, sigh.
```
What about `io.ReadFull(f)`?

timothyb89 replied:
I think `io.ReadAll()` is even better. Honestly perplexed how my googling failed to find that one, I guess I'm losing my touch.

zmb3 commented on tool/tbot/botfs/fs_linux.go:
```diff
@@ -0,0 +1,454 @@
+		})
+	}
+
+	desiredACL = append(desiredACL, []acl.Entry{
```
No need for the slice literal, right?

```
foo = append(foo, []int{1, 2, 3}...)
```

can be simplified to

```
foo = append(foo, 1, 2, 3)
```

timothyb89 replied:
I thought that ended up being syntactically nicer but it turns out I was just being dumb. Will fix.

zmb3 commented on tool/tbot/config/destination_directory.go:
```diff
@@ -58,23 +61,143 @@ func (dd *DestinationDirectory) CheckAndSetDefaults() error {
+	stat, err := os.Stat(dd.Path)
+	if trace.IsNotFound(err) {
+		err = os.MkdirAll(dd.Path, botfs.DefaultDirMode)
+		log.Infof("Created directory %q", dd.Path)
```
maybe log after the error check so you don't log if something failed?

timothyb89 replied:
Wow, how'd I miss that one? Thanks!

zmb3 commented on tool/tbot/main.go:
```diff
@@ -271,7 +373,12 @@ func getIdentityFromToken(cfg *config.BotConfig) (*identity.Identity, error) {
 		return nil, trace.Wrap(err)
 	}
-	ident, err := identity.ReadIdentityFromKeyPair(tlsPrivateKey, sshPublicKey, certs)
+	tokenHash := fmt.Sprintf("%x", sha256.Sum256([]byte(params.Token)))
```
```suggestion
	tokenHash := hex.EncodeToString(sha256.Sum256([]byte(params.Token)))
```

zmb3 commented on tool/tbot/main.go:
```diff
@@ -387,13 +544,18 @@ func renewLoop(ctx context.Context, cfg *config.BotConfig, client auth.ClientI,
 	ticker := time.NewTicker(cfg.RenewInterval)
 	defer ticker.Stop()
 	for {
+		// Make sure we can still write to the bot's destination.
```
I had to scroll down a long way to see the ticker read. Maybe it would be worth extracting the body of this loop to a separate func, so it's easy to see that renew loop is just orchestrating some function on each tick.

timothyb89 replied:
I'll pull that into its own function.

timothyb89 committed:
Fix missing error case in VerifyACL causing unreadable directories

timothyb89 commented on tool/tbot/botfs/botfs.go:
```diff
@@ -0,0 +1,156 @@
+	ACLTry ACLMode = "try"
+
+	// ACLOn enables ACL support and fails if ACLs are unavailable.
+	ACLOn ACLMode = "on"
```
Yes, that's much more clear.

timothyb89 commented on tool/tbot/botfs/botfs.go:
```diff
@@ -0,0 +1,156 @@
+// supported on all platforms and will return a trace.NotImplemented in that
+// case.
+func GetOwner(fileInfo fs.FileInfo) (*user.User, error) {
+	if runtime.GOOS == constants.WindowsOS {
```
Sure, we can probably rely on that.

timothyb89 committed:
Address review feedback

- Rename ACLOn -> ACLRequired
- Simplify fs_linux.Read()
- Add missing fs_other.Read()
- Hoist renewal loop logic into its own function
- A few misc bugfixes

rosstimothy commented on tool/tbot/init.go:
```diff
@@ -0,0 +1,534 @@
+
+	// ... and warn about / remove any unneeded files.
+	if len(toRemove) > 0 && cf.Clean {
+		log.Infof("To remove: %v", toRemove)
```
```suggestion
		log.Infof("Attempting to remove: %v", toRemove)
```

rosstimothy commented on tool/tbot/init.go:
```diff
@@ -0,0 +1,534 @@
+
+		for key, isDir := range toCreate {
+			path := filepath.Join(destDir.Path, key)
+			if err := botfs.Create(path, isDir, destDir.Symlinks); err != nil {
+				return trace.Wrap(err)
+			}
```
If this fails should we clean up any files already created?

timothyb89 replied:
It's idempotent, so I think the workflow for failure here is to correct the issue and re-run anyway. Cleaning up in general is a little bit scary (we only do it with `--clean` right now) and likely to fail anyway if a create just failed.

rosstimothy commented on tool/tbot/init.go:
```diff
@@ -0,0 +1,534 @@
+		for key := range toRemove {
+			path := filepath.Join(destDir.Path, key)
+
+			if err := os.RemoveAll(path); err != nil {
+				return trace.Wrap(err)
+			}
```
Should we attempt to remove all files and aggregate any errors instead of bailing on the first failure?

timothyb89 replied:
I like that, fixed.

timothyb89 commented on tool/tbot/botfs/botfs.go:
```diff
@@ -0,0 +1,154 @@
+
+// ACLOptions contains parameters needed to configure ACLs
+type ACLOptions struct {
+	// botUser is the bot user that should have write access to this entry
```
I really need to get a linter set up for this... thanks!

timothyb89 committed:
Apply suggestions from code review

Co-authored-by: rosstimothy <39066650+rosstimothy@users.noreply.github.com>

timothyb89 committed:
Address review feedback

- Only log syscall warning once
- Formatting and wording changes
- Improve error handling for `--clean`

timothyb89 committed:
Fix lint error

zmb3 commented on tool/tbot/config/destination_directory.go:
```diff
@@ -58,23 +61,142 @@ func (dd *DestinationDirectory) CheckAndSetDefaults() error {
+	aggregate := trace.NewAggregate(errors...)
+	if dd.ACLs == botfs.ACLRequired {
+		// Hard fail if ACLs are specifically requested and there are errors.
+		return aggregate
```
`aggregate` could be nil here though, right?

I'm wondering if you want to move the `Warnf` below because it looks like there's at least one possible code path that could exit with error and not log.

timothyb89 replied:
The thinking is that it will log the error by virtue of exiting with it (assuming there are any errors).

timothyb89 committed:
Merge branch 'master' into timothyb89/tbot-init

timothyb89 committed:
Fix imports in fs_other

r0mant commented on tool/tbot/config/destination_directory.go:
```diff
@@ -58,23 +61,142 @@ func (dd *DestinationDirectory) CheckAndSetDefaults() error {
+		return trace.Wrap(err)
+	}
+
+	switch dd.Symlinks {
```
Nit: I'd consider implementing CheckAndSetDefaults on `SymlinksMode` and `ACLMode` that do this.

timothyb89 committed:
Fix possible nil pointer deref if storage is unset

r0mant commented on tool/tbot/identity/artifact.go:
```diff
@@ -0,0 +1,143 @@
+	"github.com/gravitational/teleport/api/client/proto"
+)
+
+type Artifact struct {
```
Nit: Add godoc on the struct and fields.

r0mant commented on tool/tbot/identity/identity.go:
```diff
@@ -57,8 +56,12 @@ const (
 
-	// MetadataKey is the name under which additional metadata exists in a destination.
-	MetadataKey = "meta"
+	// TokenHashKey is the storage where a hash of the onboarding token will be stored.
```
"storage" == "directory"? I'd clarify in the comment.

r0mant commented on tool/tbot/main.go:
```diff
@@ -221,6 +251,81 @@ func onStart(botConfig *config.BotConfig) error {
+	now := time.Now().UTC()
+	if now.After(validBefore) {
+		log.Errorf(
+			"Identity appears to have expired. The renewal is likely to fail. (expires: %s, current time: %s)",
```
Nit: I think it's pretty definitive, no? :)
```suggestion
			"Identity has expired. The renewal will fail. (expires: %s, current time: %s)",
```

timothyb89 committed:
Use the bot user as default owner

This is more likely to be a safe owner choice than `nobody:nobody`.

r0mant reviewed (approved):
lgtm if you've made sure this builds on macOS/Windows (do we build tbot in Windows?)

timothyb89 committed:
Apply suggestions from code review

Co-authored-by: Roman Tkachenko <roman@goteleport.com>

timothyb89 committed:
Code review fixes

timothyb89 committed:
Merge branch 'master' into timothyb89/tbot-init

timothyb89 committed:
Merge branch 'master' into timothyb89/tbot-init
